lib/Sema/PreCheckExpr.cpp

```diff
@@ -2177,6 +2178,9 @@ bool ConstraintSystem::preCheckExpression(Expr *&expr, DeclContext *dc,
   auto &ctx = dc->getASTContext();
   FrontendStatsTracer StatsTracer(ctx.Stats, "precheck-expr", expr);
 
+  // We need to forbid from walking into anything besides closures in
```

Comment:
I think this is the other way around - `PreCheckExpression` should not walk into closures if we're delaying it until constraints are generated for each element in the body. You should be able to do that in `PreCheckExpression::walkToClosureExprPre`.

Reply:
Yes, that's is correct. I realized that this actually would create a slight problem for CSGen because `visitClosureExpr` [has to examine the body of a closure](https://github.com/apple/swift/blob/main/lib/Sema/CSGen.cpp#L2766-L2772) to determine whether it uses any type variables from outer contexts (which connects inner closure to the outer one in that case), but I think we could still do that even when UnresolvedDeclRefExpr left in the AST.